twilight-gateway/src/shard.rs

```diff
@@ -1219,70 +717,245 @@ impl<Q: Queue> Shard<Q> {
+                        self.inflater.reset();
+                    }
+                    Err(source) => {
+                        self.resume_url = None;
+                        self.state = ShardState::Disconnected {
+                            reconnect_attempts: reconnect_attempts + 1,
+                        };
+                        return Poll::Ready(Some(Err(ReceiveMessageError {
+                            kind: ReceiveMessageErrorType::Reconnect,
+                            source: Some(Box::new(source)),
+                        })));
+                    }
+                }
+            }
+            _ => {}
+        }
+
+        let message = loop {
+            ready!(self.poll_flush_pending(cx))?;
+
+            if !self.state.is_disconnected() {
+                if let Poll::Ready(frame) = self.user_channel.close_rx.poll_recv(cx) {
+                    let frame = frame.expect("shard owns channel");
+                    tracing::debug!("sending close frame from user channel");
+                    self.disconnect(CloseInitiator::Shard(frame.code));
+                    self.pending = Pending::close(Some(frame));
+                    ready!(self.poll_flush_pending(cx))?;
+                }
+            }
 
-        assert_eq!(
-            non_fatal_status,
-            ConnectionStatus::Disconnected {
-                close_code: Some(non_fatal_code),
-                reconnect_attempts: 0
+            if self
+                .heartbeat_interval
+                .as_mut()
+                .map_or(false, |heartbeater| heartbeater.poll_tick(cx).is_ready())
+            {
+                // Discord never responded after the last heartbeat, connection
+                // is failed or "zombied", see
+                // https://discord.com/developers/docs/topics/gateway#heartbeat-interval-example-heartbeat-ack
+                // Note that unlike documented *any* event is okay; it does not
+                // have to be a heartbeat ACK.
+                if self.latency.sent().is_some() && !self.heartbeat_interval_event {
+                    tracing::info!("connection is failed or \"zombied\"");
+                    self.disconnect(CloseInitiator::Shard(CloseFrame::RESUME.code));
+                    self.pending = Pending::close(Some(CloseFrame::RESUME));
+                } else {
+                    tracing::debug!("sending heartbeat");
+                    self.pending = Pending::text(
+                        json::to_string(&Heartbeat::new(self.session().map(Session::sequence)))
+                            .unwrap(),
+                        true,
+                    );
+                    self.heartbeat_interval_event = false;
+                }
+                ready!(self.poll_flush_pending(cx))?;
             }
-        );
 
-        let fatal_code = CloseCode::AuthenticationFailed;
-        let fatal_status = ConnectionStatus::from_close_code(Some(fatal_code as u16));
+            let ratelimited = self
```

Comment:
since you use `!ratelimited` later I'd suggest inverting this to `is_ready`

Reply:
`is_ready` is somewhat vague about what exactly is ready. I haven't found a good antonym for `ratelimited` and inverting it feel very clear and concise to me.

Reply:
I was referring to `is_ready` instead of `is_pending`. For the variable name I'm not too sure either, but even `not_ratelimited` would be okayish I think.

Reply:
Sorry, GitHub left me confused by not rendering enough context.